pkg/controller/certificates/duplicatesecrets/duplicatesecrets_controller.go

```diff
@@ -0,0 +1,244 @@
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		log.V(logf.DebugLevel).Info("certificate not found for key", "error", err.Error())
+		return nil
+	}
+
+	log = logf.WithResource(log, crt)
+	ctx = logf.NewContext(ctx, log)
+
+	// Get the Certificates in the same Namespace which have the same Secret name
+	// set.
+	duplicates, err := certificates.DuplicateCertificateSecretNames(ctx, c.certificateLister, crt)
+	if err != nil {
+		return err
+	}
+
+	// If we don't need to update the status, return early.
+	if !needsUpdate(crt, duplicates) {
+		return nil
+	}
+
+	condition := buildDuplicateSecretNameCondition(duplicates)
+	crt = crt.DeepCopy()
+	if condition != nil {
+		apiutil.SetCertificateCondition(crt, crt.Generation, condition.Type, condition.Status, condition.Reason, condition.Message)
+	} else {
+		// If the condition is nil, we need to remove the condition from the
+		// Certificate if it exists.
+		for i := range crt.Status.Conditions {
+			if crt.Status.Conditions[i].Type == cmapi.CertificateConditionDuplicateSecretName {
+				crt.Status.Conditions = append(crt.Status.Conditions[:i], crt.Status.Conditions[i+1:]...)
+				break
+			}
+		}
+	}
+
+	if err := c.updateOrApplyStatus(ctx, crt); err != nil {
+		return err
+	}
+
+	// We also need to re-reconcile all other Certificates which have the same
+	// Secret name set so their conditions can be evaluated.
+	for _, duplicate := range duplicates {
+		c.queue.Add(crt.Namespace + "/" + duplicate)
+	}
+
+	return nil
+}
+
+// needsUpdate returns true if the Certificate needs a state change because the
+// Certificate doesn't have a DuplicateSecretName condition when it should, or
+// that condition does not reflect the state of the world.
+func needsUpdate(crt *cmapi.Certificate, duplicates []string) bool {
+	existingCondition := apiutil.GetCertificateCondition(crt, cmapi.CertificateConditionDuplicateSecretName)
+	if len(duplicates) == 0 && existingCondition == nil {
+		return false
+	}
+
+	condition := buildDuplicateSecretNameCondition(duplicates)
+
+	// Check so that the switch below doesn't panic.
+	if existingCondition == nil && condition == nil {
+		return false
+	}
+
+	switch {
+	case (existingCondition != nil) != (condition != nil),
+		existingCondition.Status != condition.Status,
+		existingCondition.Reason != condition.Reason,
+		existingCondition.Message != condition.Message,
+		existingCondition.ObservedGeneration != crt.Generation:
+		return true
+	default:
+		return false
+	}
+}
+
+// buildDuplicateSecretNameCondition returns a nil condition if there is no
+// duplicates, else, returns a DuplicateSecretName condition.
+func buildDuplicateSecretNameCondition(duplicates []string) *cmapi.CertificateCondition {
+	if len(duplicates) == 0 {
+		return nil
+	}
+
+	const msg = "Certificate shares the same Secret name as the following Certificates in the same Namespace: [ %s ]. Issuance will block until this is resolved to prevent CertificateRequest creation runaway."
```

Comment:
Perhaps:
```suggestion
	const msg = "Certificate shares the same Secret name as the following Certificates in this Namespace: [ %s ]. Issuance will block until this is resolved to prevent CertificateRequest creation runaway."
```